lib/elixir/pages/unicode-security.md

```diff
@@ -14,12 +14,35 @@ Elixir will not allow tokenization of identifiers with codepoints in `\p{Identif
+
+We use the means described in Section 4, 'Confusable Detection', with one noted modification
+
+> Alternatively, it shall declare that it uses a modification, and provide a precise list of character mappings that are added to or removed from the provided ones.
+
+Elixir will not warn on confusability for identifiers made up exclusively of characters in a-z, A-Z, 0-9, and _. This is because ASCII identifiers have existed for so long that the programming community has had their own means of dealing with confusability between identifiers like `l,1` or `O,0` (for instance, fonts designed for programming usually make it easy to differentiate between those characters).
+
+## C3. Mixed script detection
+
+Elixir will warn on identifiers that contain a mix of characters from different scripts, but only if all of the following conditions are met:
+
+1. when those scripts are not normally used together in a writing system, as defined by Unicode's definition of 'augmented script set' and 'resolved script set'
+2. when usage of that script is comprised solely of characters that are 'mixed script confusable', per Unicode's definition of that term (confusable with characters in other scripts)
+
+The following will NOT trigger mixed-script confusable warnings:
+
+* Some languages naturally use multiple scripts. For instance, the Japanese writing system may use multiple scripts, like Hiragana, Katakana, and Han -- so an identifier in Elixir could be comprised of characters from all of those scripts (as well as Common characters, like _ and 0-9; see below).
+
+* Some letters may be used in multiple writing systems; for instance, a codepoint could appear in scripts used in the Japanese, Korean, and Chinese writing systems.
+
+* Some characters are in use in so many writing systems that they have been classified by Unicode as 'Common' or 'Inherited'; these include things like numbers, underscores, etc; Elixir will not warn about mixing of ALL-script characters, like `幻ㄒㄧㄤ1 = :foo; 幻ㄒㄧㄤ2 = :bar`.
+
+However, there are some script combinations with no overlap in characters, like {Cyrillic} and {Latin} -- in Unicode terms, the 'resolved script set' would be empty. So if that kind of script mixing occurs in an identifier, and the only Cyrillic characters in the file are those confusable with characters in other languages, it will emit a warning to that effect. (If, however, the file also contains non-confusable Cyrillic characters in source code, then the programmer can visually detect that another script is being used, and no warning is issued).
```

Comment:
Hrm... I am thinking about this... if we only emit warnings when all usages are confusable, doesn't it mean we already get the confusable warning? If so, what is the advantage of this additional rule?

Reply:
Maybe we should focus on single-script, mixed-script, and whole-script confusables for this PR, and see what C3 brings us further down the road?

I am actually keen on tightening this more than the rule above. For example, I would warn if any mixed script outside of ASCII is used. But then I would rather merge C2 first and then have this discussion.

Sorry for the back and forth on this, but those layers are being uncovered as I go deeper in the reviews (and as I re-read UTS 39. this was the third time :D).

Reply:
Yeah I wondered that at first too -- they often overlap, but not always; this handles cases where there is no other token in the file whose skeleton clashes; in that case, this would still be suspicious. (Example, may be a bit contrived): a package/dependency upstream has had an additional, visually identical function added to it; or where this is the means used to define something in another file, and it's used in the file being considered here.

In that case, usage of `is_admin()` may only happen once, so it's not actually confusable with anything in the file, but if it includes only confusable characters from Cyrillic mixed in there -- that's a fishy-enough situation to warn on (and is the example they show in Trojan Source, the sayHello() fn from a dependency, where the H is from another script). (I actually renamed the lint in this PR -- `MixedScriptUsingOnlyConfusableCharacters` -- to try to make that a bit clearer. But as with a lot of things in this standard, it can only get so clear :( because for instance 'mixed script' intuitively meant something *quite* different to me before actually implementing the standard, or rather now I have a couple of things in mind that someone could mean when they say that).

Reply:
If I got it right: C2 warns for existing confusables. C3 warns for potential confusables. Is that it?

Reply:
@mrluc if we go with a more restrictive approach, as described here:

```
Highly Restrictive

    The string qualifies as Single Script, or
    The string is covered by any of the following sets of scripts, according to the definition in Section 5.1:
        Latin + Han + Hiragana + Katakana; or equivalently: Latn + Jpan
        Latin + Han + Bopomofo; or equivalently: Latn + Hanb
        Latin + Han + Hangul; or equivalently: Latn + Kore
```

Would that simplify the implementation or make it more complex?

Reply:
Technically, C2 is about "Confusable detection" using the skeleton approach, and C3 is about "Mixed-Script detection", using the Script Resolution approach -- which is needed even to define of 'mixed script' and 'single script', since their definition is 'does/doesn't resolve to empty set'.

Then what we do with those detection capabilities is our deal; with confusables we turn them straight into warnings; with mixed-script detection we need to apply a heuristic to narrow it to genuinely suspicious cases (in both cases the same approach Rust takes).

Our confusable detection (C2) works for all 3 of those kinds of confusability, yes, by virtue of the skeleton algo, and doesn't need the C3 capabilities to do so.

Our mixed-script-with-only-confusable-characters lint, however, is a different thing -- it does require the C3 capabilities, in order to answer whether or not script mixing is occurring, but then it does the second pass to say "well, are these used in only invisible-to-programmer ways, or is the programmer aware that writing systems are mixed in their code?"

Reply:
> I am actually keen on tightening this more than the rule above. For example, I would warn if any mixed script outside of ASCII is used. But then I would rather merge C2 first and then have this discussion.
> 
> Sorry for the back and forth on this, but those layers are being uncovered as I go deeper in the reviews (and as I re-read UTS 39. this was the third time :D).

Haha, no worries, I definitely understand that feeling.

I mean I kinda love UTS39 now -- I love that they've done all that legwork to define, basically, what valid writing systems there are, and how to detect what those are (or if there isn't one) for a string, without excluding anyone! Really really love it. But yeah, stuff in the standard like "Single Script ... can be multiple scripts. Just not MIXED scripts, lol" bears a bit of digesting! KINDA wish they'd come up with a different term, like 'single writing system' and 'mixed writing system'.

And there's basic stuff in the reference impl, like using a bitmap to represent scriptsets, where I can imagine you might have ideas on how to do it cleaner, so it might be easier if C3 capabilities/lints were taken out.

(On the other hand, I don't want to send you down all the same rabbit holes, kinda worry that I'm wasting your time. But it's interesting stuff ...)

Let me know if you want this PR to go ahead with just C2, the skeleton-related confusability detection.

Reply:
> Yeah, I am on the same page now, thanks! Now my question is: why not ship a more strict version of the warning? Such as only allowing Latin with Japn, Hanb, and Kore?

Want to clarify -- do you mean the standard's recommendation of 'Highly Restrictive', or just filtering it down to only those languages?

I don't think the standard recommends *only* allowing those, notice that it mentions "single script", which requires full scriptset resolution to determine:

> Highly Restrictive:
> The **string qualifies as Single Script**, or
> The string is covered by any of the following sets of scripts, according to the definition in Section 5.1:
> Latin + Han + Hiragana + Katakana; or equivalently: Latn + Jpan
> Latin + Han + Bopomofo; or equivalently: Latn + Hanb
> Latin + Han + Hangul; or equivalently: Latn + Kore

So to determine if it's 'single script', defined in the standard, we need script resolution.

Regarding whether or not we could do something more restrictive -- YES, that's a lang design question for sure.

* For instance, I initially thought "single script is good enough for identifiers"; it ignores Common and all mathy chars are Common; it resolves multiple-script asian languages, etc.

* The counter-example there would be variables written by people in other languages, who are now forced to write "twitter" and "API" in their script 🤦.

However, how much simpler does it make things? Well, since resolution returns the resolved set, and the full set, it could basically remove the fns `check_confusable_mixed_script` and `warning_per_problematic_script`.

But I think almost everything we do with mixed-script would scriptset resolution, which is the bulk of the C3 additions.

Reply:
@josevalim github UI lag got me on [your question](https://github.com/elixir-lang/elixir/pull/11582#discussion_r786882791) here; I responded to it above -- "Single Script" is the key there; it would simplify it somewhat, removing a couple of functions, but not the whole script resolution process.

Reply:
@mrluc I see. Let's go with C2 for now only. Then, after merging, I can split it into a separate module, which should leave us in a better position to accept C3. WDYT?

Reply:
Sounds good to me!

After talking over the C3 stuff, C2 seems *blessedly* simple! 😆 Since it's relatively simpler it may as well be the only thing in the PR. I'll probably be able to do that + incorporate the feedback tomorrow.

Reply:
Thanks, and again, sorry for the false positive on accepting both C2 and C3 at once. :)